tools/webdriver/webdriver/bidi/client.py

```diff
@@ -0,0 +1,393 @@
+        """Create a BiDi session where there is no existing HTTP session
+
+        :param webdocket_url: URL to the WebSocket server listening for BiDi connections
+        :param requested_capabilities: Capabilities request for establishing the session."""
+        return cls(websocket_url, requested_capabilities=requested_capabilities, loop=loop)
+
+    async def __aenter__(self) -> "BidiSession":
+        await self.start()
+        return self
+
+    async def __aexit__(self, *args: Any) -> None:
+        await self.end()
+
+    async def start(self) -> None:
+        """Connect to the WebDriver BiDi remote via WebSockets"""
+        self.transport = Transport(self.websocket_url, self.on_message, loop=self.loop)
+
+        if self.session_id is None:
+            self.capabilities = await self.session.new(self.requested_capabilities)
+
+        await self.transport.start()
+
+    async def send_command(self, method: str, params: Mapping[str, Any]) -> Awaitable[Mapping[str, Any]]:
+        """Send a command to the remote server"""
+        # this isn't threadsafe
+        self.command_id += 1
+        command_id = self.command_id
+
+        body = {
+            "id": command_id,
+            "method": method,
+            "params": params
+        }
+        assert command_id not in self.pending_commands
+        self.pending_commands[command_id] = self.loop.create_future()
+        assert self.transport is not None
+        await self.transport.send(body)
+
+        return self.pending_commands[command_id]
+
+    async def on_message(self, data: Mapping[str, Any]) -> None:
+        """Handle a message from the remote server"""
+        if "id" in data:
+            # This is a command response or error
+            future = self.pending_commands.get(data["id"])
+            if future is None:
+                raise ValueError(f"No pending command with id {data['id']}")
+            if "result" in data:
+                future.set_result(data["result"])
+            elif "error" in data and "message" in data:
+                assert isinstance(data["error"], str)
+                assert isinstance(data["message"], str)
+                future.set_exception(BidiException(data["error"],
+                                                   data["message"],
+                                                   data.get("stacktrace")))
+            else:
+                raise ValueError(f"Unexpected message: {data!r}")
+        elif "method" in data and "params" in data:
+            # This is an event
+            method = data["method"]
+            listeners = self.event_listeners.get(method, [])
+            if not listeners:
+                listeners = self.event_listeners.get(None, [])
+            for listener in listeners:
+                await listener(method, data["params"])
+        else:
+            raise ValueError(f"Unexpected message: {data!r}")
```

Comment:
Is it intended the method ignores other than expected fields? I'd assume the concept "ignore everything unknown" allows quicker and easier test writings, but there can be some side effects with unexpected fields. The same belongs to the principle of the subscribing to events.

Reply:
Allowing additional fields is intentional; that's the approach the spec takes to forward compatability. See https://w3c.github.io/webdriver-bidi/#protocol-definition ; for the top-level fields we have the `*test => any` definition (and I think the idea is that any complex object will ignore unknown fields in the same way. If we want things to be more tightly defined we'd need to open an issue on the spec itself (although the argument for strict schemas is strong when you control both ends of the wire, it's less obvious when there's an loosly coupled ecosystem of multiple idepednent clients and servers; I think if we want to tighten things up we'd need a story for transitioning clients/servers to future versions of the protocol when it may not be fully implemented everywhere).

Reply:
Thanks for additional context, it was what I was asking about.

Reply:
The `*test => any` bits in the spec indeed mean that we can't assert that there's nothing beyond "id", "method" and "params" here without going further than the spec does. But I also think it would make sense to do both of these things:
- Require that browsers (remote ends) ignore extra keys when interpreting JSON objects as commands
- Test that browsers *not* send any extra top-level keys with events they emit

If we change the protocol, the tests can change.

I'm not sure I'd make the same argument for parameters (keys inside "params") however, since they're much more likely to change over time, and dealing with those test failures when adding experimental (off-by-default) support for new event parameters would be a nuisance, probably.

Reply:
I've filed https://github.com/w3c/webdriver-bidi/issues/111, let's not include it in this PR.